### PR TITLE
Initial codelens implementation for C#

### PR DESCRIPTION
**Pull request by awschristou**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [ ] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [ ] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description

Initial implementation adding C# codelenses.
There is a portion that searches csproj files for an assembly name. This will be done in a separate change.

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/codelens/csharpCodeLensProvider.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+async function getAssemblyName(sourceCodeUri: vscode.Uri): Promise<string | undefined> {
+    const projectFile: vscode.Uri | undefined = await findParentProjectFile(sourceCodeUri)
+
+    if (!projectFile) {
+        return undefined
+    }
+
+    // TODO : Perform an XPATH parse on the project file
+    // If Project/PropertyGroup/AssemblyName exists, use that. Otherwise use the file name.
```
Comment (mpiroc) [inline src/shared/codelens/csharpCodeLensProvider.ts]: What if the `AssemblyName` is specified in a `.targets`/`.props` file, and does not match the filename?

File: src/shared/codelens/csharpCodeLensProvider.ts [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+
+            return sourceCodeUri.fsPath.startsWith(dirname)
+        })
+        .sort()
+        .reverse()
+
+    return parentProjectFiles.length === 0 ? undefined : parentProjectFiles[0]
+}
+
+export function isPublicClassSymbol(
```
Comment (mpiroc) [inline src/shared/codelens/csharpCodeLensProvider.ts]: `s/method/class/g`

File: src/shared/codelens/csharpCodeLensProvider.ts [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    document: Pick<vscode.TextDocument, 'getText'>,
+    symbol: vscode.DocumentSymbol,
+): boolean {
+    if (symbol.kind === vscode.SymbolKind.Class) {
+        const methodText: string = document.getText(symbol.range)
```
Comment (mpiroc) [inline src/shared/codelens/csharpCodeLensProvider.ts]: What is `symbol.selectionRange` in this case? Is it just `MyClass`, or is it `public class MyClass`?

File: src/shared/codelens/csharpCodeLensProvider.ts [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+        return METHOD_SYMBOL_PUBLIC_REGEXP.test(signatureBeforeMethodName)
+    }
+
+    return false
+}
+
+export function getMethodNameFromSymbol(symbol: vscode.DocumentSymbol): string | undefined {
+    if (symbol.kind === vscode.SymbolKind.Method) {
+        const matches = symbol.name.match(METHOD_SYMBOL_NAME_REGEXP)
```
Comment (mpiroc) [inline src/shared/codelens/csharpCodeLensProvider.ts]: In what case will a `SymbolKind.Method` symbol not match this regex? Just if the user hasn't typed it yet?

File: src/shared/codelens/csharpCodeLensProvider.ts (hunk 3 above)
Comment (mpiroc) [reply src/shared/codelens/csharpCodeLensProvider.ts]: Also, what if the user has non-compiling code like:

```csharp
private Foo {
    Console.WriteLine("publicclass")
```

File: src/test/shared/codelens/csharpCodeLensProvider.test.ts [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+const fakeRange = new vscode.Range(0, 0, 0, 0)
+
+describe('findParentProjectFile', async () => {
+    const sourceCodeUri = vscode.Uri.file(path.join('code', 'someproject', 'src', 'Program.cs'))
+    const projectInSameFolderUri = vscode.Uri.file(path.join('code', 'someproject', 'src', 'App.csproj'))
+    const projectInParentFolderUri = vscode.Uri.file(path.join('code', 'someproject', 'App.csproj'))
+    const projectInParentParentFolderUri = vscode.Uri.file(path.join('code', 'App.csproj'))
+    const projectOutOfParentChainUri = vscode.Uri.file(path.join('code', 'someotherproject', 'App.csproj'))
+
+    const testScenarios = [
```
Comment (micholas) [inline src/test/shared/codelens/csharpCodeLensProvider.test.ts]: Nice :)

File: src/shared/codelens/codeLensUtils.ts [hunk 6]
```diff
@@ -113,10 +113,10 @@ function makeConfigureCodeLens({
     samTemplate
 }: MakeConfigureCodeLensParams): vscode.CodeLens {
     // Handler will be the fully-qualified name, so we also allow '.' despite it being forbidden in handler names.
-    if (/[^\w\-\.]/.test(handlerName)) {
+    if (/[^\w\-\.\:]/.test(handlerName)) {
         throw new Error(
             `Invalid handler name: '${handlerName}'. ` +
-            'Handler names can contain only letters, numbers, hyphens, and underscores.'
+            'Handler names can contain only letters, numbers, hyphens, underscores, and colons.'
```
[the remark this answers is not in the document]
Comment (awschristou) [reply src/shared/codelens/codeLensUtils.ts]: Handler names are runtime specific. For example, [dotnet](https://docs.aws.amazon.com/lambda/latest/dg/dotnet-programming-model-handler-types.html#dotnet-programming-model-handler-signatures) allows `:` and `.`

This message should only surface when someone is developing this extension, and has done something wrong. We should not need such a detailed message. I'll scale it back.

File: src/shared/codelens/csharpCodeLensProvider.ts [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    const logger = getLogger()
+
+    const codeLensProvider: vscode.CodeLensProvider = {
+        provideCodeLenses: async (
+            document: vscode.TextDocument,
+            token: vscode.CancellationToken
+        ): Promise<vscode.CodeLens[]> => {
+            const handlers: LambdaHandlerCandidate[] = await getLambdaHandlerCandidates(document)
+            logger.debug(
+                'csharpCodeLensProvider.makePythonCodeLensProvider handlers:',
```
[the remark this answers is not in the document]
Comment (awschristou) [reply src/shared/codelens/csharpCodeLensProvider.ts]: busted

File: src/shared/codelens/csharpCodeLensProvider.ts [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+    const symbols: vscode.DocumentSymbol[] = (
+        (await vscode.commands.executeCommand<vscode.DocumentSymbol[]>(
+            'vscode.executeDocumentSymbolProvider',
+            document.uri
+        )) || []
+    )
+
+    return getLambdaHandlerSymbolsTuplets(document, symbols)
+        .map<LambdaHandlerCandidate>(tuplet => {
+            const handlerName = produceHandlerName(assemblyName, tuplet)
```
[the remark this answers is not in the document]
Comment (awschristou) [reply src/shared/codelens/csharpCodeLensProvider.ts]: code is the music of my life

File: src/shared/codelens/csharpCodeLensProvider.ts (hunk 3 above)
Comment (awschristou) [reply src/shared/codelens/csharpCodeLensProvider.ts]: A Symbol wouldn't be emitted

File: src/shared/codelens/csharpCodeLensProvider.ts [hunk 9]
```diff
@@ -32,13 +32,15 @@ export const CSHARP_ALLFILES: vscode.DocumentFilter[] = [
...
 ]
 
-const METHOD_SYMBOL_NAME_REGEXP = new RegExp(/\w*/)
-const METHOD_SYMBOL_PUBLIC_REGEXP = new RegExp(/\bpublic\b/)
-
-export interface DotNetHandlerSymbolsTuplet {
-    namespace: vscode.DocumentSymbol,
-    class: vscode.DocumentSymbol,
-    method: vscode.DocumentSymbol,
+const REGEXP_RESERVED_WORD_PUBLIC = new RegExp(/\bpublic\b/)
```
Comment (mpiroc) [inline src/shared/codelens/csharpCodeLensProvider.ts]: Nit: You can just use a regex literal:

```typescript
const REGEXP_RESERVED_WORD_PUBLIC = /\bpublic\b/
```

File: src/shared/codelens/csharpCodeLensProvider.ts [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+
+    return path.parse(projectFile.fsPath).name
+}
+
+export function getLambdaHandlerComponents(
+    document: vscode.TextDocument,
+    symbols: vscode.DocumentSymbol[],
+    assembly: string,
+): DotNetLambdaHandlerComponents[] {
+    return symbols
```
Comment (mpiroc) [inline src/shared/codelens/csharpCodeLensProvider.ts]: What if the class is not in a namespace. Is this a valid scenario?

Comment (awschristou) [reply src/shared/codelens/csharpCodeLensProvider.ts]: I don't believe it is. We should address that on an as-needed basis.

**Pull request merged**